Agregada petición API para panel de control

joaquinperez94 opened the pull request:
**¿Que hace este pull request?**
Se han implementado los cambios necesarios para que la API de almacenamiento de votos,sea capaz de devolver los votantes de una votación pasada por parámetro.

[Link de la incidencia](https://github.com/danhidsan/decide-ganimedes-almacenamiento/issues/3)

**¿Por que estas haciendo el pull request?. Contexto del pull request.**
Esta funcionalidad se ha implementado con el fin de que sea usada en el desarrollo del panel de control y mediante una llamada a la API se sepa quién ha votado en la votación que se pasa por parámetro.

**¿Por dónde debe empezar el revisor?**
Deberá de crear al menos una votación y un par de usuarios que voten dicha votación.
Para probar la petición deberemos de obtener el id de la votación que hemos creado.
Una vez que tenemos el id iniciamos el servidor con:

`python manage.py runserver`

Una vez iniciado, deberemos de poner en el navegador la siguiente url, donde id_voting hace referencia al id de la votación que creamos anteriormente:

`http://localhost:8000/store/votings/**id_voting**/`

Dicha url nos deberá devolver un json en el que aparecerán los dos votantes que votaron en dicha votación. Algo como así, que nos muestran 3 usuarios que votaron en la votación que le pasamos por parámetro

{
"User": [
{
"id": 2,
"last_login": null,
"is_superuser": false,
"username": "julia",
"first_name": "",
"last_name": "",
"email": "",
"is_staff": false,
"is_active": true,
"date_joined": "2018-12-03T17:36:31.945Z"
},
{
"id": 3,
"last_login": null,
"is_superuser": false,
"username": "maria",
"first_name": "",
"last_name": "",
"email": "",
"is_staff": false,
"is_active": true,
"date_joined": "2018-12-03T17:36:46.067Z"
},
{
"id": 4,
"last_login": null,
"is_superuser": false,
"username": "pepe",
"first_name": "",
"last_name": "",
"email": "",
"is_staff": false,
"is_active": true,
"date_joined": "2018-12-03T17:41:16.735Z"
}

joaquinperez94 committed:
Agregada petición API para panel de control

danhidsan commented on decide/store/views.py:
```diff
@@ -6,9 +6,23 @@
+    def get(self,request,voting_id):
+        voting_id = self.kwargs['voting_id']
+        result=Vote.objects.filter(voting_id=voting_id).values('voter_id')
+        #for id in result
```
Este código comentado hay que borrarlo. Ensucia nuestro código y por tanto el repositorio común.

danhidsan commented on decide/store/views.py:
```diff
@@ -6,9 +6,23 @@
+        #    result=User.objects.filter(id=id).values('username')
+        usernames = User.objects.filter(id__in=result).values('id','last_login','is_superuser','username','first_name','last_name',
+                                                              'email','is_staff','is_active','date_joined')
+        #return JsonResponse({"Vote": list(result)})
```
Este código comentado tambien hay que borrarlo

danhidsan commented on decide/store/views.py:
```diff
@@ -6,9 +6,23 @@
+        usernames = User.objects.filter(id__in=result).values('id','last_login','is_superuser','username','first_name','last_name',
+                                                              'email','is_staff','is_active','date_joined')
+        #return JsonResponse({"Vote": list(result)})
+        return JsonResponse({"User": list(usernames)})
```
No comprendo, porque usamos un "JsonResponse" en el código de más abajo hay un ejemplo de cómo devolver la respuesta con Django REST.

danhidsan reviewed (changes requested):
El funcionamiento de la incidencia es correcto. Pero hay varias cosas que hay que corregir:

1. El estilo del código: Hay código comentado, lo cual no es una buena práctica, ya que ensucia el código. Ademas no se siguen las reglas de estilo de Python, habiendo lineas que superan el numero mínimo de letras por linea.
2. No se utiliza las "Response" de Django REST: Con el fin de seguir el mismo patrón con el código ya existente, deberíamos utilizar las misma librerías que ya se usan si no hay causa de fuerza mayor.
3. El formato de la respuesta no es del todo correcto: Ahora mismo se devuelve un objeto que dentro tiene un array de objetos, el objeto "User". Esto no es necesario, es posible devolver el array de objetos directamente.
4. En la respuesta no se devuelve una "status": Es necesario que el servidor cada vez que devuelva algo, tenga que devolver el estado de la respuesta. Ej. En caso de que la respuesta sea buena, tendrá que devolver un código 200.

julgargal commented:
> **¿Por dónde debe empezar el revisor?**
> Deberá de crear al menos una votación y un par de usuarios que voten dicha votación.
> Para probar la petición deberemos de obtener el id de la votación que hemos creado.

Para comprobar el buen funcionamiento del pull request y validarlo ¿podrías detallar como obtener el id de la votación creada?

joaquinperez94 committed:
style:Modificación código requeridos para aprobar pull request

joaquinperez94 commented:
Ya están hechos los cambios y se puede volver a revisar el pull request.
Además se ha añadido la restricción que sólo podrán realizar la petición aquellos usuarios que sean staff.

danhidsan committed:
style: Cambiado el formato de la url para obtener el los votantes de una votación

Se ha cambiado el formato de la url para que tenga una nomenglatura mas clara para su uso.

Incidencia: #3

danhidsan commented:
@julgargal Para obtener el id de la votación hay que acceder a la interfaz de administración de decide, ahí se puede obtener todos los IDs de todas las entradas de la base de datos.

danhidsan commented:
Se ha cambiado el formato de la URL de la petición para que sea mas clara para la persona que vaya a hacer uso de ella. Pero la funcionalidad es correcta y funciona de manera adecuada.
